### PR TITLE
Revert "[19.09] Fix detached parent instance error in 8769"

### PR DESCRIPTION
**Pull request by bgruening**
I think we need to revert this one. It increases the memory usage of handler 4 fold.

grafik

handler3 is running the reverted Galaxy codebase. Handler 0,1,2 are running this patch. As you can see the memory consumption is 4 times higher :(

Sorry!

### CODE REVIEW

**Comment by martenson** [comment]
@bgruening thanks for production-testing this. Unfortunate results though :/

**Comment by dannon** [comment]
Oh wow, that's a huge bump in memory usage.  I'm glad you noticed that so quickly.  @mvdbeek Do we just need to enforce refreshing the objects at each access to solve the original issue?

**Comment by mvdbeek** [comment]
you mean [refresh](https://docs.sqlalchemy.org/en/13/orm/session_api.html#sqlalchemy.orm.session.Session.refresh) ? I don't think that would work but I haven't tried.

At this point there are a couple of questions:
- How is it possible that eagerloading a couple of thousand sqlalchemy objects would consume so much memory?
- Could it be that 20G is the "normal" load (possibly caused by another issue), and handler3 is just not doing any heavy work (because it can't access some tool attributes without raising the DetachedInstanceError) ? 

I don't have access to a server with as many repos as usegalaxy.eu, but I don't see any difference in memory consumption with 557 repos.

The root cause of the problem seems to be that the cache is established in a sqlalchemy session that gets closed at one point, which leads to the `DetachedInstanceError`, see https://docs.sqlalchemy.org/en/13/errors.html#error-bhk3. So the PR we're reverting here is attempting point 2 in the mitigation strategies. We could try giving the ToolShedRepositoryCache its own session though, that should probably work as well ?

**Comment by bgruening** [comment]
> Could it be that 20G is the "normal" load (possibly caused by another issue), and handler3 is just not doing any heavy work (because it can't access some tool attributes without raising the DetachedInstanceError) ?

Unfortunately, not all handlers are processing. And 20GB was not the normal load we had a cgroup memory cap on the processes and recognized after applying the patch that we needed to increase the memory cap from 15GB (which was already way too much) to 25 GB to not constantly restart handlers.
So there is a steep increase in usage from what I can tell.

**Comment by martenson** [comment]
xref https://github.com/galaxyproject/galaxy/pull/8780

**Comment by martenson** [comment]
I am testing this on Test, do not backport yet please

**Comment by martenson** [comment]
some data from test.galaxyproject.org

* without https://github.com/galaxyproject/galaxy/pull/8780
Screen Shot 2019-10-11 at 11 58 04 AM

* with https://github.com/galaxyproject/galaxy/pull/8780 (shortly after restart, server up)
Screen Shot 2019-10-11 at 1 40 20 PM


TLDR I do not see confirmation of what @bgruening reported with regards to memory. In my understanding the cache would bloat the memory during startup, which is in both these cases finished.

**Comment by bgruening** [comment]
We do see memory increasing all the time, even after startup :(
I'm not sure what else it can be.

Without the patch we are back to normal levels:
```
root@sn04:~$ gxadmin uwsgi memory
galaxy-zergpool.service 3325952
galaxy-zergling@2.service 6842474496
galaxy-zergling@1.service 9766760448
galaxy-zergling@0.service 12578988032
galaxy-handler@7.service 5239414784
galaxy-handler@0.service 5523763200
galaxy-handler@1.service 5280866304
galaxy-handler@5.service 5319712768
galaxy-handler@2.service 5371699200
galaxy-handler@3.service 5290250240
galaxy-handler@4.service 5156126720
galaxy-handler@6.service 5189091328

root@sn04:~$ head -n 100 /opt/galaxy/server/lib/galaxy/model/tool_shed_install/mapping.py | tail -n 20
                                                   Column("tool_id", Integer, ForeignKey("tool_version.id"), index=True, nullable=False),
                                                   Column("parent_id", Integer, ForeignKey("tool_version.id"), index=True, nullable=False))

install_model.MigrateTools.table = Table("migrate_tools", metadata,
                                         Column("repository_id", TrimmedString(255)),
                                         Column("repository_path", TEXT),
                                         Column("version", Integer))

mapper(install_model.ToolShedRepository, install_model.ToolShedRepository.table,
       properties=dict(tool_versions=relation(install_model.ToolVersion,
                                              primaryjoin=(install_model.ToolShedRepository.table.c.id == install_model.ToolVersion.table.c.tool_shed_repository_id),
                                              backref='tool_shed_repository'),
                       tool_dependencies=relation(install_model.ToolDependency,
                                                  primaryjoin=(install_model.ToolShedRepository.table.c.id == install_model.ToolDependency.table.c.tool_shed_repository_id),
                                                  order_by=install_model.ToolDependency.table.c.name,
                                                  backref='tool_shed_repository'),
                       required_repositories=relation(install_model.RepositoryRepositoryDependencyAssociation,
                                                      primaryjoin=(install_model.ToolShedRepository.table.c.id == install_model.RepositoryRepositoryDependencyAssociation.table.c.tool_shed_repository_id))))

mapper(install_model.RepositoryRepositoryDependencyAssociation, install_model.RepositoryRepositoryDependencyAssociation.table,


```

**Comment by martenson** [comment]
fwiw these are main handlers which would more closely relate to yours, this should be with https://github.com/galaxyproject/galaxy/pull/8780 in 
marten_galaxy-web-04____ssh_

**Comment by bgruening** [comment]
I applied the patch again. Its not that bad as yesterday (might be due to less traffic), but:

>    Active: active (running) since Sat 2019-10-12 08:36:06 CEST; 1h 22min ago
   Active: active (running) since Sat 2019-10-12 06:00:45 CEST; 3h 58min ago
   Active: active (running) since Sat 2019-10-12 04:28:17 CEST; 5h 30min ago
   Active: active (running) since Sat 2019-10-12 07:47:31 CEST; 2h 11min ago
   Active: active (running) since Sat 2019-10-12 09:29:21 CEST; 29min ago
   Active: active (running) since Sat 2019-10-12 04:45:17 CEST; 5h 13min ago
   Active: active (running) since Sat 2019-10-12 09:41:45 CEST; 17min ago
   Active: active (running) since Sat 2019-10-12 07:13:36 CEST; 2h 45min ago

and 

> root@sn04:~$ journalctl -u galaxy-h* --since "10 hours ago" | grep KILL
Oct 12 00:25:08 sn04.bi.uni-freiburg.de systemd[1]: galaxy-handler@2.service: main process exited, code=killed, status=9/KILL
Oct 12 01:10:04 sn04.bi.uni-freiburg.de systemd[1]: galaxy-handler@1.service: main process exited, code=killed, status=9/KILL
Oct 12 01:26:02 sn04.bi.uni-freiburg.de systemd[1]: galaxy-handler@0.service: main process exited, code=killed, status=9/KILL
Oct 12 02:06:34 sn04.bi.uni-freiburg.de systemd[1]: galaxy-handler@7.service: main process exited, code=killed, status=9/KILL
Oct 12 02:06:52 sn04.bi.uni-freiburg.de systemd[1]: galaxy-handler@2.service: main process exited, code=killed, status=9/KILL
Oct 12 02:07:09 sn04.bi.uni-freiburg.de systemd[1]: galaxy-handler@3.service: main process exited, code=killed, status=9/KILL
Oct 12 02:15:41 sn04.bi.uni-freiburg.de systemd[1]: galaxy-handler@4.service: main process exited, code=killed, status=9/KILL
Oct 12 02:19:38 sn04.bi.uni-freiburg.de systemd[1]: galaxy-handler@5.service: main process exited, code=killed, status=9/KILL
Oct 12 02:34:27 sn04.bi.uni-freiburg.de systemd[1]: galaxy-handler@3.service: main process exited, code=killed, status=9/KILL
Oct 12 03:42:22 sn04.bi.uni-freiburg.de systemd[1]: galaxy-handler@5.service: main process exited, code=killed, status=9/KILL
Oct 12 03:55:16 sn04.bi.uni-freiburg.de systemd[1]: galaxy-handler@1.service: main process exited, code=killed, status=9/KILL
Oct 12 03:57:03 sn04.bi.uni-freiburg.de systemd[1]: galaxy-handler@5.service: main process exited, code=killed, status=9/KILL
Oct 12 04:00:53 sn04.bi.uni-freiburg.de systemd[1]: galaxy-handler@4.service: main process exited, code=killed, status=9/KILL
Oct 12 04:11:32 sn04.bi.uni-freiburg.de systemd[1]: galaxy-handler@5.service: main process exited, code=killed, status=9/KILL
Oct 12 04:23:31 sn04.bi.uni-freiburg.de systemd[1]: galaxy-handler@2.service: main process exited, code=killed, status=9/KILL
Oct 12 04:28:17 sn04.bi.uni-freiburg.de systemd[1]: galaxy-handler@5.service: main process exited, code=killed, status=9/KILL
Oct 12 04:45:17 sn04.bi.uni-freiburg.de systemd[1]: galaxy-handler@3.service: main process exited, code=killed, status=9/KILL
Oct 12 06:00:44 sn04.bi.uni-freiburg.de systemd[1]: galaxy-handler@0.service: main process exited, code=killed, status=9/KILL
Oct 12 06:00:44 sn04.bi.uni-freiburg.de systemd[1]: galaxy-handler@6.service: main process exited, code=killed, status=9/KILL
Oct 12 07:13:36 sn04.bi.uni-freiburg.de systemd[1]: galaxy-handler@4.service: main process exited, code=killed, status=9/KILL
Oct 12 07:47:30 sn04.bi.uni-freiburg.de systemd[1]: galaxy-handler@7.service: main process exited, code=killed, status=9/KILL
Oct 12 08:36:06 sn04.bi.uni-freiburg.de systemd[1]: galaxy-handler@1.service: main process exited, code=killed, status=9/KILL
Oct 12 09:29:21 sn04.bi.uni-freiburg.de systemd[1]: galaxy-handler@2.service: main process exited, code=killed, status=9/KILL
Oct 12 09:41:45 sn04.bi.uni-freiburg.de systemd[1]: galaxy-handler@6.service: main process exited, code=killed, status=9/KILL


This indicates that during job submission / tool view we are leaking memory, isn't it?

> head -n 100 /opt/galaxy/server/lib/galaxy/model/tool_shed_install/mapping.py | tail -n 20
                                                   Column("tool_id", Integer, ForeignKey("tool_version.id"), index=True, nullable=False),
                                                   Column("parent_id", Integer, ForeignKey("tool_version.id"), index=True, nullable=False))

install_model.MigrateTools.table = Table("migrate_tools", metadata,
                                         Column("repository_id", TrimmedString(255)),
                                         Column("repository_path", TEXT),
                                         Column("version", Integer))

mapper(install_model.ToolShedRepository, install_model.ToolShedRepository.table,
       properties=dict(tool_versions=relation(install_model.ToolVersion,
                                              primaryjoin=(install_model.ToolShedRepository.table.c.id == install_model.ToolVersion.table.c.tool_shed_repository_id),
                                              backref='tool_shed_repository',
                                              lazy='joined'),
                       tool_dependencies=relation(install_model.ToolDependency,
                                                  primaryjoin=(install_model.ToolShedRepository.table.c.id == install_model.ToolDependency.table.c.tool_shed_repository_id),
                                                  order_by=install_model.ToolDependency.table.c.name,
                                                  backref='tool_shed_repository',
                                                  lazy='joined'),
                       required_repositories=relation(install_model.RepositoryRepositoryDependencyAssociation,
                                                      primaryjoin=(install_model.ToolShedRepository.table.c.id == install_model.RepositoryRepositoryDependencyAssociation.table.c.tool_shed_repository_id),


Patch is still included. @martenson, @mvdbeek I guess you need to create traffic so see this behaviour.

**Comment by bgruening** [comment]
Another indicator:

```
root@sn04:~$ systemctl status galaxy-h* | grep Memory
   Memory: 10.8G (limit: 20.0G)
   Memory: 11.2G (limit: 20.0G)
   Memory: 12.7G (limit: 20.0G)
   Memory: 10.8G (limit: 20.0G)
   Memory: 5.9G (limit: 20.0G)
   Memory: 12.5G (limit: 20.0G)
   Memory: 4.1G (limit: 20.0G)
   Memory: 4.1G (limit: 20.0G)
root@sn04:~$ systemctl status galaxy-h* | grep ago
   Active: active (running) since Sat 2019-10-12 08:36:06 CEST; 1h 39min ago
   Active: active (running) since Sat 2019-10-12 06:00:45 CEST; 4h 15min ago
   Active: active (running) since Sat 2019-10-12 04:28:17 CEST; 5h 47min ago
   Active: active (running) since Sat 2019-10-12 07:47:31 CEST; 2h 28min ago
   Active: active (running) since Sat 2019-10-12 09:29:21 CEST; 46min ago
   Active: active (running) since Sat 2019-10-12 04:45:17 CEST; 5h 30min ago
   Active: active (running) since Sat 2019-10-12 09:41:45 CEST; 34min ago
   Active: active (running) since Sat 2019-10-12 10:03:28 CEST; 12min ago
```

You see nicely that the memory usage correlates with the uptime.

**Comment by bgruening** [comment]
It getting morning here in Europe. After 500+ jobs it looks worse. I'm going to revert again.

```
root@sn04:/opt/galaxy/server/tools/interactive$ systemctl status galaxy-h* | grep Memory
   Memory: 3.3G (limit: 20.0G)
   Memory: 3.6G (limit: 20.0G)
   Memory: 9.3G (limit: 20.0G)
   Memory: 18.2G (limit: 20.0G)
   Memory: 18.6G (limit: 20.0G)
   Memory: 19.0G (limit: 20.0G)
   Memory: 3.6G (limit: 20.0G)
   Memory: 3.5G (limit: 20.0G)
root@sn04:/opt/galaxy/server/tools/interactive$ systemctl status galaxy-h* | grep ago
   Active: active (running) since Sat 2019-10-12 12:05:13 CEST; 5min ago
   Active: active (running) since Sat 2019-10-12 12:01:13 CEST; 9min ago
   Active: active (running) since Sat 2019-10-12 11:48:49 CEST; 21min ago
   Active: active (running) since Sat 2019-10-12 11:48:25 CEST; 21min ago
   Active: active (running) since Sat 2019-10-12 11:47:54 CEST; 22min ago
   Active: active (running) since Sat 2019-10-12 11:47:20 CEST; 23min ago
   Active: active (running) since Sat 2019-10-12 12:01:53 CEST; 8min ago
   Active: active (running) since Sat 2019-10-12 12:03:01 CEST; 7min ago
```

**Comment by hexylena** [comment]
It's a huge difference:

Current status:

image

Previously Normal:

image

**Comment by mvdbeek** [comment]
So I think I figured out one way we can "leak" memory here ... I don't think we need a thread local cache here anymore. I'll re-open this without `self.cache = threading.locals()`. And also on closer investigation the cache looked kind-of broken, probably because of this.
